Feature/sqlalchemy dataset

jpcampb2 opened the pull request:
This PR creates an initial SqlAlchemyDataSet with the following implemented expectations. 
See #12 
 - `expect_column_to_exist`
 - `expect_table_row_count_to_be`
 - `expect_table_row_count_to_be_between`
 - `expect_column_values_to_not_be_null`
 - `expect_column_values_to_be_null`
 - `expect_column_values_to_be_in_set`
 - `expect_column_values_to_be_between`
 - `expect_column_mean_to_be`
 - `expect_column_min_to_be`
 - `expect_column_max_to_be`
 - `expect_column_sum_to_be`

The PR also introduces the new system of testing expectations for many expectations to ensure as much parity across systems as possible. See #167

jcampbell committed:
Move pandas distributional expectations to new submodule

jcampbell committed:
Allow _format_column_map_output to properly handle None for index list.

jcampbell committed:
Add SqlAlchemyDataSet

jcampbell committed:
Add fixture-ized test for column_map_expectation

jcampbell committed:
Add pytest requirement to dev environment

jcampbell committed:
Initial menu of test running configurations using fixtures...

jcampbell committed:
Update parameterization to identify individual tests.

jcampbell committed:
Refactor json_test_runner to modularize.

jcampbell committed:
Update travis to use pytest

jcampbell committed:
Add sqlalchemy to travis install script for testing.

jcampbell committed:
Remove explicit import in tests module.

jcampbell committed:
Tweak json test evaluation setup.

jcampbell committed:
Clarify test runner setup.
 - Note that we are now less efficient in reusing datasets built via parameterization.

jcampbell committed:
Additional variations to discuss...

jcampbell committed:
Add additional option for running tests.

jcampbell committed:
Add additional sqlalchemy_dataset expectations

jcampbell committed:
Continue to update test development model

jcampbell committed:
Updates to util function for evaluating the json test type.

abegong committed:
Done?

jcampbell committed:
Add concept of excepted logic to evaluate_json_test

abegong committed:
Decruft

abegong committed:
Remove now-extraneous test files

abegong committed:
Clean up expect_column_values_to_be_between.json: titles and out fields

abegong committed:
Moar decruft

abegong committed:
Move test_set files around

abegong committed:
Set up expectation_types and datasets for all column_map_expectations.

abegong committed:
expectation_type and datasets for other_expectations

abegong committed:
expectation_type and datasets for column_aggregate_expectations

abegong committed:
Titles from all other_expectations

abegong committed:
Titles for column_aggregate_expectations

abegong committed:
Titles for column_map_expectations

abegong committed:
exact_map_out for column_map_expectations

abegong committed:
exact_map_out for other_expectations

abegong committed:
exact_map_out for column_aggregate_expectations

abegong committed:
Add out fields

abegong committed:
Fix expect_table_row_count_to_be_between_test_set.json

abegong committed:
Drop *_test_set suffixes from testing files.

abegong committed:
Remove refs to result_obj

abegong committed:
Remove json-ported tests from test_pandas_dataset.py

abegong committed:
Clean up test_great_expectations.py

abegong committed:
Write test_is_on_temporary_notimplemented_list check

abegong committed:
Fix misnamed expectations

abegong committed:
Rename other_expectation json files

abegong committed:
Remove array check for observed_value; Remove extraneous terms in column_aggregate_expectation checks

jcampbell committed:
Tweak testing framework to not generate tests for not-implemented conditions, and to raise a warning in such cases.

jcampbell committed:
Add known exception for cross-type-comparisons in SqlAlchemy

jcampbell committed:
Refactor approach for SqlAlchemyDataSet to build query in expectation but execute in decorator for simpler codepath and potential future optimizations (or dry run)

jcampbell committed:
Refactor sqlalchemydataset to reduce number of queries for column_map_expectations

jcampbell committed:
Small syntax improvement to sqlalchemydataset column_map_expectation

jcampbell committed:
Remove old implementation style

jcampbell committed:
Add helper to parse result format.

jcampbell committed:
Add test for expect_table_row_count_to_equal and sqlalchemy implementation

jcampbell committed:
refactor tests/util to tests/test_utils

jcampbell committed:
Add sqlalchemy implementation and test for column_exist and table_row_count_equal

jcampbell committed:
Update all test runners to not generate tests on notimplemented list

jcampbell committed:
Add SqlAlchemy column max

jcampbell committed:
Add column_aggregate_expectation decorator for SqlAlchemy

jcampbell committed:
Add sqlalchemy column_max expectation and test

jcampbell committed:
Add column_sum expectation and test for sqlalchemy

jcampbell committed:
Remove ties_okay parameter erroneously listed in expect_column_sum_to_be_between expectation

jcampbell committed:
Add expect_column_mean_to_be_between to SqlAlchemy and associated tests.

jcampbell committed:
Correct test spec files.

jcampbell committed:
Add SqlAlchemy support for column-index parameter

jcampbell committed:
Add default column existence expectations for SqlAlchemyDataSet

abegong commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -0,0 +1,496 @@
+
+    @classmethod
+    def column_map_expectation(cls, func):
+        """For SqlAlchemy, this decorator allows individual column_map_expectations to simply return the filter
```
💯

abegong commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -0,0 +1,496 @@
+
+            unexpected_condition = func(self, column, *args, **kwargs)
+
+            count_query = select([
```
Nice consolidation!

abegong commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -0,0 +1,496 @@
+
+            count_results = self.engine.execute(count_query).fetchone()
+
+            unexpected_query_results = self.engine.execute(
```
This is very cool. I hadn't realized that we would be able to be so direct.

abegong commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -0,0 +1,496 @@
+            if ('success' not in evaluation_result) or \
+                ('result_obj' not in evaluation_result) or \
+                ('observed_value' not in evaluation_result['result_obj']):
+                raise ValueError("Column aggregate expectation failed to return required return information.")
```
Minor nit: this error message could be more informative...

abegong replied:
(Fixing this probably doesn't belong on this PR.)

abegong commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -0,0 +1,496 @@
+                }
+            })
+
+    def _is_numeric_column(self, column):
```
?

abegong replied:
hm. We've brought typing in via the backdoor.

jpcampb2 replied:
True. I'd argue it's no more severe than what happens right now with the implied type coercions in python (that we inherit in PandasDataSet). Actually, this was necessary to ensure we're staying closer between the two right now. But I'm definitely looking forward to a more elegant solution, e.g. as implied in #207 .

abegong replied:
:+1:

abegong commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -0,0 +1,496 @@
+            raise ValueError("parse_strings_as_datetimes is not supported in SqlAlchemy")
+
+        col_max = self.engine.execute(
+            select([sa_func.max(sa_column(column))]).select_from(table(self.table_name))
```
Got it. So `column_aggregate_expectations` act almost exactly like they do in `PandasDataSet`.

Side thought: we need to update our docs for custom expectations.

jpcampb2 replied:
Good catch. I updated docs for custom expectations and also added an example of a custom SqlAlchemyDataSet that uses both kinds of decorators.

abegong commented on great_expectations/dataset/util.py:
```diff
@@ -1,18 +1,25 @@
 from functools import wraps
 
 
+def parse_result_format(result_format):
```
? A docstring would be helpful here.

jpcampb2 replied:
Done.

abegong commented on tests/column_aggregate_expectations/test_column_aggregate_expectations.py:
```diff
@@ -0,0 +1,64 @@
+###
+###
+#
+# This file should not be modified. To adjust test cases, edit the related json file.
```
"file(s)"

jpcampb2 replied:
👍 :)

abegong commented on tests/test_pandas_dataset.py:
```diff
@@ -10,108 +10,26 @@
-            D = ge.dataset.PandasDataSet(J["dataset"])
-            D.set_default_expectation_argument("result_format", "COMPLETE")
-            T = J["tests"]
+    # def test_expect_column_to_exist(self):
```
We might as well decruft all these commented tests, don't you think?

jpcampb2 replied:
Agreed. Pulled them out.

abegong reviewed (changes requested):
This is awesome work. One of my favorite PRs so far.

I have a few small changes to request, but on the whole, this set of changes feels really, really good.

jcampbell committed:
Add check on value of min_value and max_value for expect_column_mean_to_be_between

jcampbell committed:
Provide more informative error messages for missing elements from column_aggregate_expectation-decorated functions

jcampbell committed:
Clean up sqlalchemy imports for more consistent naming.

jcampbell committed:
Clarify logic of MetaSqlAlchemyDataSet by inverting returned condition from column_map_expectation to be the expected condition rather than the unexpected condition.

jcampbell committed:
Add documentation and test for Custom SqlAlchemyDataSet objects

jcampbell committed:
Update json_test_runner template docs and initialize_expectations in each one.

jcampbell committed:
Remove tests that have been migrated to the new format.

jcampbell committed:
Add support for missing min/max in expect_column_values_to_be_between in SqlAlchemy
 - And add missing min/max to tests (thank you coverage for spotting that for us!)

jcampbell committed:
Coverage hunting: missing min/max parameters in expect_column_mean_to_be_between

jcampbell committed:
Add docstring for parse_result_format and use it more consistently.

jcampbell committed:
Update changelog with info on PR

jcampbell committed:
Add sqlalchemy to requirements

jcampbell committed:
Change SqlAlchemyDataSet constructor to accept either connection string or engine.

jcampbell committed:
Refine handling of partial counts in SqlAlchemyDataSet to limit results unless result_obj_format is COMPLETE

abegong commented on great_expectations/dataset/sqlalchemy_dataset.py:
```diff
@@ -297,7 +293,7 @@ def expect_column_values_to_be_null(self,
     ):
 
-        return sa_column(column) != None
+        return sa.column(column) == None
```
Yikes. Good catch.

abegong reviewed (approved):
This looks great!

jcampbell committed:
Add additional testing for SqlAlchemyDataSet expectations

jpcampb2 commented:
Woohoo!

abegong commented:
:tada::tada::tada:!
